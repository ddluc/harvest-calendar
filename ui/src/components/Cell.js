import React from 'react';

function Cell ({mPos, cellIndex, id}) {

  if (mPos.status === 'none' || mPos.status === 'off') {
    return (<div id={id} className={`calendar-cell cell-${cellIndex}`}>—</div>);
  } else if (mPos.status === 'early') {
    return (<div id={id} className={`calendar-cell cell-${cellIndex}`}>E</div>);
  } else if (mPos.status === 'peak') {
    return (<div id={id} className={`calendar-cell cell-${cellIndex}`}>P</div>);
  } else if (mPos.status === 'late') {
    return (<div id={id} className={`calendar-cell cell-${cellIndex}`}>L</div>);
  } else {
    return (<div className={`calendar-cell`}>—</div>);
  }

};

export default Cell;
