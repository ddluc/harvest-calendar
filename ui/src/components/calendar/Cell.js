import React from 'react';
import { Store } from '../../flux/store';

function Cell ({mPos, cellIndex, id, isActive}) {

  const { state, dispatch } = React.useContext(Store);

  let cellContent = '';

  if (state.pricingMode == 'on') {
    cellContent = mPos.value > 0 ? `${mPos.value}` : '-';
  } else if (mPos.status === 'none' || mPos.status === 'off') {
    cellContent = '-';
  } else if (mPos.status === 'early') {
    cellContent = 'E';
  } else if (mPos.status === 'peak') {
    cellContent = 'P'
  } else if (mPos.status === 'late') {
    cellContent = 'L'
  }

  return (<div id={id} className={`calendar-cell cell-${cellIndex} ${isActive} ${mPos.status}`} >{cellContent}</div>);

};

export default Cell;
