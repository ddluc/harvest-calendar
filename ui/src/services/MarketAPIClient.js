import axios from 'axios';

class MarketAPIClient {

  constructor() {
    this.baseUrl = 'http://localhost:3300';
  }

  async getMarketData () {
    const url = `${this.baseUrl}/api/market_data`;
    try {
      const response = await axios.get(url);
      return response.data; 
    } catch (err) {
      console.error(err);
      return null;
    }
  }

}

export default new MarketAPIClient();
