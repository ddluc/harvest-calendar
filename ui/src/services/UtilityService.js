/**
 * Utility Methods
 * (i.e) Selectors
 **/

/**
 * getMonthsInSeason
 * @description gets the months for the active season
 * @param <Array> months — array of months
 * @param <Number> season — the id of the active season (1,2,3,4)
 */
export const getMonthsInSeason = (months, season) => {
  let monthsInSeason = [];
  months.forEach((month) => {
    if (month.season.includes(season)) {
      monthsInSeason.push(month);
    }
  });
  return monthsInSeason;
};

/**
 * getVegetablesInSeason
 * @description gets the vegetables for the active season
 * @param <Array> vegetables - array of vegetable objects
 * @param <Array> months — array of months
 * @param <Number> season — the id of the active season (1,2,3,4)
 */
export const getVegetablesInSeason = (vegetables, months, season) => {
  let vegetablesInSeason = [];
  const monthsInSeason = getMonthsInSeason(months, season);
  const monthKeys = monthsInSeason.map((month) => month.key);
  vegetables.forEach((vegetable) => {
    let inSeason = false;
    vegetable.mPos.forEach((mPos) => {
      if (monthKeys.includes(mPos.month)){
        if (mPos.status !== 'none') {
          inSeason = true;
        }
      }
    });
    if (inSeason) {
      vegetablesInSeason.push({ ...vegetable})
    }
  });
  vegetablesInSeason.sort((a, b) => b.mPosCount - a.mPosCount);
  return vegetablesInSeason;
};

/**
 * sortVegetablesByScore
 * @description sorts the vegetables by custom algorithm to determine how in season it is
 * @param <Array> vegetables - array of vegetable objects
 * @param <Array> months — array of months
 * @param <Number> season — the id of the active season (1,2,3,4)
 */
export const sortVegetables = (vegetables, months, season) => {
  let vegetablesInSeason = [];
  const monthsInSeason = getMonthsInSeason(months, season);
  const monthKeys = monthsInSeason.map((month) => month.key);
  vegetables.forEach((vegetable) => {
    let inSeason = false;
    vegetable.mPos.forEach((mPos) => {
      if (monthKeys.includes(mPos.month)){
        if (mPos.status !== 'none') {
          inSeason = true;
        }
      }
    });
    if (inSeason) {
      vegetablesInSeason.push({ ...vegetable})
    }
  });
  vegetablesInSeason.sort((a, b) => b.mPosCount - a.mPosCount);
  return vegetablesInSeason;
};

/**
 * buildCellMap
 * @description maps css grid cell indexes to month/segment
 * @param <Array> months — array of months
 * @param <Number> season — the id of the active season (1,2,3,4)
 * @returns <Object> cellMap — a map of cell indexes
 */
export const buildCellMap = (months, season) => {
  let cellMap = {body: {}, header: {}};
  let bodyCellIndex = 1;
  let headerCellIndex = 1;
  let segments = buildSegmentMap();
  let monthsInSeason = getMonthsInSeason(months, season);
  monthsInSeason.forEach((month) => {
    cellMap.header[month.key] = headerCellIndex;
    headerCellIndex++;
    segments.forEach((segment) => {
      let key = `${month.key}-${segment.id}`;
      cellMap.body[key] = bodyCellIndex;
      bodyCellIndex++;
    });
  });
  return cellMap;
};


/**
 * buildSegmentMap
 * @description builds a statuc segment array
 * @returns <Array> segment
 */
export const buildSegmentMap = () => {
  const segments = [
    {id: 1, key: 'early', label: 'Early'},
    {id: 2, key: 'mid', label: 'Mid', },
    {id: 3, key: 'late', label: 'Late'}
  ];
  return segments;
};
