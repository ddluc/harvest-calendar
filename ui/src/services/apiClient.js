/**
 * API Client
 */

import axios from 'axios';

const getBaseURL = () => {
  let baseURL = '';
  if (document.location.hostname === 'localhost') {
    baseURL = 'http://localhost:3300';
  } else {
    baseURL = 'http://harvestapp.ddluc.com';
  }
  return baseURL;
}

export const getMarketData = async () => {
  const url = `${getBaseURL()}/api/market_data`;
  try {
    const response = await axios.get(url);
    return response.data;
  } catch (err) {
    console.error(err);
    return null;
  }
}
