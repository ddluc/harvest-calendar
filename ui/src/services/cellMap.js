/**
 * Maps the market data to the grid cells
 */

export const buildCellMap = (activeSeason) => {

  let months = [];
  switch (activeSeason) {
    case 1:
      months = ['MAR', 'APR', 'MAY', 'JUN'];
      break;
    case 2:
      months = ['JUN', 'JUL', 'AUG', 'SEP'];
      break;
    case 3:
      months = ['SEP', 'OCT', 'NOV', 'DEC'];
      break;
    case 4:
      months = ['DEC', 'JAN', 'FEB', 'MAR'];
      break;
    default:
      months = ['MAR', 'APR', 'MAY', 'JUN'];
  }

  let cellMap = {body: {}, header: {}};
  let bodyCellIndex = 1;
  let headerCellIndex = 1;
  let segments = buildSegmentMap();
  months.forEach((month) => {
    cellMap.header[month] = headerCellIndex;
    headerCellIndex++;
    segments.forEach((segment) => {
      let key = `${month}-${segment.id}`;
      cellMap.body[key] = bodyCellIndex;
      bodyCellIndex++;
    });
   });
  return cellMap;

}

export const buildSegmentMap = () => {
  const segments = [
    {id: 1, key: 'early', label: 'Early'},
    {id: 2, key: 'mid', label: 'Mid', },
    {id: 3, key: 'late', label: 'Late'}
  ];
  return segments;
};
