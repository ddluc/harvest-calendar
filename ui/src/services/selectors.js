/**
 * Utility Methods
 **/

/**
 * getMonthsInSeason
 * @description gets the months for the active season
 * @param <Object> state - application state
 */
export const getMonthsInSeason = (state) => {
  let { months } = state.marketData;
  let { activeSeason } = state;
  let monthsInSeason = [];
  let order = [];
  switch (activeSeason) {
    // Spring
    case 1:
      order = ['MAR', 'APR', 'MAY', 'JUN'];
      break;
    // Summer
    case 2:
      order = ['JUN', 'JUL', 'AUG', 'SEP'];
      break;
    // Fall
    case 3:
      order = ['SEP', 'OCT', 'NOV', 'DEC'];
      break;
    // Winter
    case 4:
      order = ['DEC', 'JAN', 'FEB', 'MAR'];
      break;
    default:
      order = ['MAR', 'APR', 'MAY', 'JUN'];
  }
  order.forEach((key) => {
    months.forEach((month) => {
      if (month.key === key) {
        monthsInSeason.push(month);
      }
    });
  });
  return monthsInSeason;
};


/**
 * sortVegetables
 * @description sorts the vegetables by custom algorithm to determine how in season it is
 * @param <Object> state - application state
 */
export const sortVegetables = (state) => {
  let {marketData, activeMonth, activeSegment} = state;
  let peak = [];
  let early = [];
  let late = [];
  let off = [];
  let none = [];
  if (activeSegment) {
    marketData.vegetables.forEach((vegetable) => {
      vegetable.mPos.forEach((mPos) => {
        if (`${mPos.month}-${mPos.segment}` === `${activeMonth}-${activeSegment}` ) {
          if ( mPos.status === 'peak') peak.push(vegetable);
          if ( mPos.status === 'early') early.push(vegetable);
          if ( mPos.status === 'late') late.push(vegetable);
          if ( mPos.status === 'off') off.push(vegetable);
          if ( mPos.status === 'none') none.push(vegetable);
        }
      });
    });
    let sortedVegetables = [...peak, ...early, ...late, ...off, ...none ];
    return sortedVegetables;
  } else {
    return marketData.vegetables;
  }
};

/**
 * filterVegetables
 * @description filters the vegetables to the user's search query
 * @param <Object> state - application state
 */
export const filterVegetables = (state, vegetables) => {
  let { searchInput } = state;
  console.log(searchInput);
  console.log(vegetables);
  if (searchInput !== '' && searchInput.length > 2) {
    let filteredVegetables = [];
    vegetables.forEach((veg) => {
      if (veg.label.includes(searchInput)) {
        filteredVegetables.push(veg);
      }
    });
    return filteredVegetables;
  } else {
    return vegetables;
  }
}

/**
 * buildCellMap
 * @description maps the market data to the css grid
 * @param <Object> state - application state
 */
export const buildCellMap = (state) => {
  const segments = buildSegmentMap();
  const months = getMonthsInSeason(state);
  let cellMap = {body: {}, header: {}};
  let bodyCellIndex = 1;
  let headerCellIndex = 1;
  months.forEach((month) => {
    cellMap.header[month.key] = headerCellIndex;
    headerCellIndex++;
    segments.forEach((segment) => {
      let key = `${month.key}-${segment.id}`;
      cellMap.body[key] = bodyCellIndex;
      bodyCellIndex++;
    });
   });
  return cellMap;
}

export const buildSegmentMap = () => {
  const segments = [
    {id: 1, key: 'early', label: 'Early'},
    {id: 2, key: 'mid', label: 'Mid', },
    {id: 3, key: 'late', label: 'Late'}
  ];
  return segments;
};
