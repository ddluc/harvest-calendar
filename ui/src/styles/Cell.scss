@import '__theme.scss';

.calendar-cell {
  text-align: center;
  padding-top: 15px;
  border-top: solid 1px $grey-4;
  border-right: solid 1px $grey-4;
}

.calendar-cell.active {
  font-weight: bold;
  color: $charcol;
  border-left: solid 2px $charcol;
  border-right: solid 2px $charcol; 
}

.calendar-cell.price{
  padding-top: 10px;
  font-size: 11px;
  &.none {
    padding-top: 15px;
    font-size: 16px;
  }
  span { display: block; }

}

.calendar-cell.early {
  background-color: $early;
}

.calendar-cell.peak {
  background-color: $peak;
}

.calendar-cell.late {
  background-color: $late;
}
