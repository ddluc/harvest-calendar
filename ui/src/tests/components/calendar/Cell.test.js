import React from 'react';
import Cell from '../../../components/calendar/Cell';
import * as selectors from '../../../services/selectors';
import { getMockState } from '../../__mocks/state.mock.js';
import {
  mockBuildSegmentMap,
  mockGetMonthsInSeason,
  mockBuildCellMap
} from '../../__mocks/selectors.mock.js';
import { shallow } from 'enzyme';

describe('Cell', () => {

  it('should render', () => {
    let mockState = getMockState();
    // Select a random vegetable to test against
    let vegetableIndex = 20;
    let vegetable = mockState.marketData.vegetables[vegetableIndex];
    let mPos = vegetable.mPos[6];
    buildContextSpy(mockState);
    const props = {
      mPos,
      cellIndex: 1,
      id: 'MAR-1',
      isActive: false,
      showPrice: false, 
      category: 'other'
    };
    let wrapper = shallow(<Cell { ...props} /> );
    expect(wrapper).toMatchSnapshot();
  })

});
