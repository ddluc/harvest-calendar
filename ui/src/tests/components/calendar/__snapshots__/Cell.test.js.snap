// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render 1`] = `
<div
  className="calendar-cell other"
  data-cell-index={1}
  data-cell-status="none"
  id="MAR-1"
/>
`;
